client/state-db/src/pruning.rs

```diff
@@ -113,36 +104,30 @@ impl<BlockHash: Hash, Key: Hash, D: MetaDb> DeathRowQueue<BlockHash, Key, D> {
 	fn new_db_backed(
 		db: D,
 		base: u64,
-		mut uncached_blocks: usize,
+		last: Option<u64>,
 		window_size: u32,
 	) -> Result<DeathRowQueue<BlockHash, Key, D>, Error<D::Error>> {
 		// limit the cache capacity from 1 to `DEFAULT_MAX_BLOCK_CONSTRAINT`
 		let cache_capacity = window_size.max(1).min(DEFAULT_MAX_BLOCK_CONSTRAINT) as usize;
 		let mut cache = VecDeque::with_capacity(cache_capacity);
 		trace!(target: "state-db", "Reading pruning journal for the database-backed queue. Pending #{}", base);
-		// Load block from db
-		DeathRowQueue::load_batch_from_db(
-			&db,
-			&mut uncached_blocks,
-			&mut cache,
-			base,
-			cache_capacity,
-		)?;
-		Ok(DeathRowQueue::DbBacked { db, cache, cache_capacity, uncached_blocks })
+		DeathRowQueue::load_batch_from_db(&db, &mut cache, base, cache_capacity)?;
+		Ok(DeathRowQueue::DbBacked { db, cache, cache_capacity, last })
 	}
 
 	/// import a new block to the back of the queue
-	fn import(&mut self, base: u64, journal_record: JournalRecord<BlockHash, Key>) {
+	fn import(&mut self, base: u64, num: u64, journal_record: JournalRecord<BlockHash, Key>) {
 		let JournalRecord { hash, inserted, deleted } = journal_record;
+		trace!(target: "state-db", "Importing {}, base={}", num, base);
 		match self {
-			DeathRowQueue::DbBacked { uncached_blocks, cache, cache_capacity, .. } => {
-				// `uncached_blocks` is zero means currently all block are loaded into `cache`
-				// thus if `cache` is not full, load the next block into `cache` too
-				if *uncached_blocks == 0 && cache.len() < *cache_capacity {
+			DeathRowQueue::DbBacked { cache, cache_capacity, last, .. } => {
+				// If the new block continues cached range and there is space, load it directly into
+				// cache.
+				if num == base + cache.len() as u64 && cache.len() < *cache_capacity {
```

Comment:
`num == base + cache.len() as u64` this should always be true or? And we just have this here to be defensive?

Reply:
Yes, normally insertions should be sequential and in order.